lib/matplotlib/image.py

```diff
@@ -420,13 +420,19 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
                     a_min, a_max = np.int32(0), np.int32(1)
                 if inp_dtype.kind == 'f':
                     scaled_dtype = A.dtype
-                    # Cast to float64
-                    if A.dtype not in (np.float32, np.float16):
-                        if A.dtype != np.float64:
-                            cbook._warn_external(
-                                f"Casting input data from '{A.dtype}' to "
-                                f"'float64' for imshow")
-                        scaled_dtype = np.float64
+                    # Cast unsupported dtypes to nearest supported float depth
+                    if A.dtype not in (np.float64, np.float32):
+                        # cast float16 up into float32
+                        if A.dtype == np.float16:
+                            scaled_dtype = np.float32
+                        # cast float128 down into float64
+                        elif A.dtype == np.float128:
```

Comment:
This misses `float96`, and I think `float128` may be optional on some platforms, so it's best to turn this into an inverted condition, like before.